src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,284 @@
+        if (index > size / 2) {
+            current = tail;
+            counter = size - 1;
+            step = -1;
+            iterationVariant = FROM_END_TYPE_OF_ITERATION;
+        }
+        while (current != null) {
+            if (counter == index) {
+                oldValue = current.value;
+                current.value = value;
+            }
+            current = iterationVariant.equals(FROM_START_TYPE_OF_ITERATION)
+                    ? current.next : current.prev;
+            counter += step;
+        }
+        return oldValue;
     }
 
     @Override
-    public T remove(int index) {
-        return null;
+    public boolean remove(T object) {
+        if (head == null || isEmpty()) {
+            return false;
+        }
+        int index = indexOf(object);
+        if (index != -1) {
+            remove(index);
+            return true;
+        } else {
+            return false;
+        }
     }
 
     @Override
-    public boolean remove(T object) {
-        return false;
+    public T remove(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(INDEX_OF_BOUND_EXCEPTION_MESSAGE);
+        }
+        T deletedValue;
+        if (size == 1) {
+            deletedValue = head.value;
+            removeLastElement();
+        } else if (index == 0) {
+            deletedValue = head.value;
+            removeElementFromBeginning();
+        } else if (index == size - 1) {
+            deletedValue = tail.value;
+            removeElementFromEnd();
+        } else {
+            deletedValue = removeFromMiddle(index);
+        }
+        size--;
+        return deletedValue;
+    }
+
+    private void removeElementFromEnd() {
+        tail = tail.prev;
+        tail.next = null;
+    }
+
+    private void removeElementFromBeginning() {
+        head = head.next;
+        head.prev = null;
+    }
+
+    private void removeLastElement() {
+        head = null;
+        tail = null;
+    }
+
+    private T removeFromMiddle(int index) {
+        Node<T> current = head;
+        int counter = 0;
+        int step = 1;
+        String iterationVariant = FROM_START_TYPE_OF_ITERATION;
+        if (index > size / 2) {
+            current = tail;
+            counter = size - 1;
+            step = -1;
+            iterationVariant = FROM_END_TYPE_OF_ITERATION;
+        }
+        while (current != null) {
+            if (counter == index) {
+                break;
+            }
+            current = iterationVariant.equals(FROM_START_TYPE_OF_ITERATION)
+                    ? current.next : current.prev;
+            counter += step;
+        }
+        unlink(current);
+        return current.value;
+    }
+
+    private void unlink(Node<T> current) {
+        if (current != null) {
+            Node<T> newNext = current.next;
+            Node<T> newPrev = current.prev;
+            newNext.prev = newPrev;
+            newPrev.next = newNext;
```

Comment:
can `newNext` or `newPrev` be null?

Reply:
rework remove methods and delete methods removeElementFromEnd, removeElementFromBeginning, removeLastElement.
Add a null-check code in unlink() method